python/paddle/distributed/passes/auto_parallel_sharding.py

```diff
@@ -0,0 +1,694 @@
+# 
+#     http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import reduce
+from collections import OrderedDict
+import numpy as np
+
+import paddle
+from paddle.framework import core
+from paddle.fluid import unique_name
+from .pass_base import PassBase, register_pass
+from paddle.distributed.fleet.meta_optimizers.common import is_backward_op, is_optimizer_op
+from paddle.distributed.auto_parallel.process_group import get_world_process_groups, new_process_group
+from paddle.distributed.auto_parallel.operators.common import is_parameter_related
+from paddle.distributed.auto_parallel.utils import _get_comm_group, naive_set_dist_op_attr_for_program_by_mesh_and_mapping, set_var_dist_attr
+
+OpRole = core.op_proto_and_checker_maker.OpRole
+OP_ROLE_KEY = core.op_proto_and_checker_maker.kOpRoleAttrName()
+_skip_ops = ['create_py_reader', 'create_double_buffer_reader', 'read', 'slice']
+# update here to support new optimizers
+_supported_optimizer_type = [
+    "adam", "adamax", "adamw", "decayed_adagrad", "momentum", "dgc_momentum",
+    "lars_momentum", "merged_momentum", "lamb", "sgd"
+]
+
+
+# NOTE we add the "auto_parallel" prefix to the pass in order to
+# indicate that this pass should obey some constrains by auto_parallel
+# for example all ops and vars should has dist attr before and after pass
+# should use dist op instead of custom comm op 
+@register_pass("auto_parallel_sharding")
+class ShardingPass(PassBase):
+    def __init__(self):
+        super(ShardingPass, self).__init__()
+        self.set_attr("dist_context", None)
+        self.set_attr("stage", None)
+        self.set_attr("sharding_degree", None)
+        self.set_attr("params_grads", [])
+        self.set_attr("global_rank", -1)
+        self.dp_groups = set()
+        self.sharding_infos = []
+        self.varname_to_sharding_info = {}
+        self.partial_sharding = False
+        self.outer_dp_group = None
+
+    def _check_self(self):
+        if self.get_attr("dist_context") is None:
+            return False
+
+        if self.get_attr("stage") not in [1, 2, 3]:
+            return False
+        if (not isinstance(self.get_attr("sharding_degree"),
+                           int)) or self.get_attr("sharding_degree") <= 1:
+            return False
+        if len(self.get_attr("params_grads")) <= 0:
+            return False
+        if (not isinstance(self.get_attr("global_rank"),
+                           int)) or self.get_attr("global_rank") < 0:
+            return False
+
+        return True
+
+    def _check_conflict(self, other_pass):
+        return True
+
+    def _apply_single_impl(self, main_program, startup_program, context):
```

Comment:
I remember that our sharding pass does not support multiple blocks. How about add some assertion here?